flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/mailbox/MailboxSender.java

```diff
@@ -0,0 +1,52 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.streaming.runtime.tasks.mailbox;
+
+import javax.annotation.Nonnull;
+
+/**
+ * Producer-facing side of the {@link Mailbox} interface. This is used to enqueue letters. Multiple producers threads
+ * can put to the same mailbox.
+ */
+public interface MailboxSender {
+
+	/**
+	 * Enqueues the given letter to the mailbox, if capacity is available. On success, this returns <code>true</code>
+	 * and <code>false</code> if the mailbox was already full.
+	 *
+	 * @param letter the letter to enqueue.
+	 * @return <code>true</code> iff successful.
+	 */
+	boolean tryPutMail(@Nonnull Runnable letter);
```

Comment:
Do we need non blocking variant here?

Reply:
I think it can be helpful if client code that wants to run code can be non-blocking if required.

Reply:
Yes sure, it can be helpful, but do we need it now? Why should we maintain and test a now unused code? We can always introduce it later and the later we introduce it, the fewer refactorings/bug fixing/maintenance time we spend on it.

Reply:
I was asking myself the same question, and here are my arguments why I decided to introduce it:
- most importantly, from all I can forsee, this interface will at some point be exposed to user code, e.g. via some source context. I think therefore it is ok to introduce things proactively if user can find a benefit in them instead of waiting for them to ask for it. And I can think of use cases where non-blocking support can be beneficial.
- symmetry in the interface between input and output.
- the test effort it took me is like adding 3 lines, the method has a low complexity. Opportunity cost to introduce it here is therefore reasonably low.

Reply:
> most importantly, from all I can forsee, this interface will at some point be exposed to user code

if that's the case I agree. Having said that, are we planning to expose it?

Reply:
Here is my line of thought: user code currently is using the checkpointing lock to execute atomic operations w.r.t. event processing. If we are planning to remove the lock, we should give users a proper substitute for the mechanism. Executing one letter can also be regarded an atomic operation w.r.t. event processing. The exposed interface would then be the `MailboxSender` interface, probably at least through whatever will be the future source contexts, maybe also in other places (sinks come to my mind). It does not take much creativity to think of cases where non-blocking options can help there as well.